### ISSUE TITLE
Custom priorities, comment support when using an internal tokenizer

### ISSUE THREAD

**Issue by nikomatsakis** [nikomatsakis/lalrpop#14]
The current tokenizer generation always uses two fixes precedence categories, so all regular expressions have equal weight. This is useful for giving a keyword like `"class"` precedence of an identifier regex, but there are times when we would like to give some regexs higher precedence over others. For example, if parsing a case-insensitive language like Pascal, you would like to use regexs like `r"[vV][aA][rR]"` rather than having to enumerate all combinations. But this won't work because of precedence conflicts. Another problem is that the tokenizer implicitly skips whitespace and there is no way to extend this set to skip other things, like comments.

### Prioritization

I've been contemplating a `match` declaration to address prioritization, which might look like:

```rust
match {
    r"[vV][aA][rR]",
    r"[a-zA-Z][a-zA-Z0-9]+",
}
```

The idea here is that when you have tokens listed in a match declaraction, we can create custom precedence levels, so that e.g. here the "var" regex takes precedence over the identifier regex. Tokens not listed would be implicitly added to the end of the list, with literals first and regex second.

### Comments and whitespace

I'm less clear on what to do here. I contemplated adding things to the match declaration with an "empty action", which would signal "do nothing":

```rust
match {
    r"{.*}" => { } // pascal style comment
}
```

or having something like an empty if declaration:

```rust
if r"{.*}";
```

I think I prefer the first alternative, but it doesn't seem great. Another thing that is unclear is if the implicit whitespace skipping should be disabled or retained. I think I prefer to retain it unless the user *asks* for it to be disabled, because it's always a bit surprising when you add something and find it implicitly removes another default. That is, adding comments to the list of things to skip implicitly removes whitespace. But not having the implicit whitespace at all feels like clearly the wrong default.

So some uncertainties here!

**Comment by nikomatsakis** [comment]
I've been thinking about this more and I am feeling pretty good about `match` declarations as the means of writing tokenizer rules. To start with I'd probably just permit:

```
match {
    <pattern> => { }
}
```

Which would indicate that matches of `<pattern>` are to be ignored completely. The priority would be given the same priorities as today (0 for regular expressions, 1 for fixed strings). 

In the future, I would expand this to permit:

1. Explicit `#[priority=N]` declarations.
2. Fallible actions (`=>?`).
3. Actions that produce a token value or multiple token values. See below.
4. States. Here I imagine `match in foo` or `match in (foo, bar)` where `foo` and `bar` are states, and some way to transition to a new state. Not quite sure how that should look; maybe even something like `self.push_state(foo)` in the action, though I'm generally not too keen on "injecting" identifiers into the action code.

The matter of actions that produce tokens still requires some thought. My initial idea was that they could return something that supports `IntoIterator<T>`, where `T` is the token type, but that would require specifying the types of tokens. Given that the current tokenizer type is just `(usize, &'input str)`, that's not so good. We'd need some way to give names to the token types that get produced. 

It might be that what you want then is a combination of an "extern token" and custom tokenizer -- basically a way to declare a token enum, and then using `match` statements to declare code that converts from text into those variants. We'd probably some nice shorthands then so that you can just list out the variants and not have to write explicit patterns for them.

**Comment by davll** [comment]
`lex` or `flex` has **start conditions** to define custom states for special cases like comments.

[flex doc](http://dinosaur.compilertools.net/flex/flex_11.html)
[Reference Link](http://stackoverflow.com/questions/2130097/difficulty-getting-c-style-comments-in-flex-lex)

**Comment by nikomatsakis** [comment]
@davll yes, I am hoping to solve that another way (https://github.com/nikomatsakis/lalrpop/issues/195). In any case I purposefully left it out of this initial design.

**Comment by nikomatsakis** [comment]
er, wrong issue, but still true :P

**Comment by divoxx** [comment]
Any progress on this? Being able to match against newline specifically would be really really welcomed. Even if the solution was simply an option to toggle multiline matching on the lexer, allowing to choose whether newline is to be matched as whitespace.

**Comment by benkay86** [comment]
Love the LALRPOP crate, but not having support for comments out of the box is a big letdown.  All of the [Example Uses in README.md](https://github.com/lalrpop/lalrpop#example-uses) have comments, including LALRPOP itself, so it almost seems hypocritical not to support them.  I realize that it's possible to implement comments by writing a custom lexer, but that's not trivial to do!

A lot of ambitious ideas are put forward in this issue, and it looks like the `match` syntax has [made it into the book](http://lalrpop.github.io/lalrpop/lexer_tutorial/001_lexer_gen.html).  If there were a way to just get the comment syntax proposed by [nikomatsakis](https://github.com/nikomatsakis) supported that would be a tremendous help!  Other more advanced features of a custom tokenizer could come later.

```
// Ignore <pattern>, a way to implement comments.
match {
    <pattern> => { }
}
```

**Comment by Marwes** [comment]
This would be great, and it is probably the most requested feature from LALRPOP. Unfortunately there isn't anyone actively maintaining the crate. I believe I am the most active member and I don't even remember to answer all issues and questions. I'd be happy to review a PR (but it may require some prodding to make me remember) but ultimately I don't, nor do I believe any other maintainer has the bandwidth to implement this.

**Comment by Marwes** [comment]
Implemented the most basic, naive version I could in #509 .

**Comment by aegooby** [comment]
The regex/literal based version is great, but it has issues with nested C-style comments like
```c
/* /* */ */
```
This syntax
```
// Ignore <pattern>, a way to implement comments.
match {
    <pattern> => { }
}
```
would be perfect for handling these, is there any implementation planned, or maybe a workaround using the current system?

**Comment by airstrike** [comment]
Curious if there's any progress on allowing for significant whitespace with LALRPOP. I tend to love the ergonomics of the crate overall and am trying not to move away from it when writing a whitespace-significant DSL

**Comment by dburgener** [comment]
The workaround described [here](https://github.com/lalrpop/lalrpop/issues/678) should allow you to parse whitespace using the built in lexer in lalrpop.
